### PR TITLE
NEP: Initial draft for NEP 43 for extensible ufuncs

### PR DESCRIPTION
**Pull request by seberg**
There is still quite a bit to be done here. However, some of the important things for discussion are there I think, so putting it out here to work on here and hope for feedback.

One of the thoughts I had with respect to https://github.com/numpy/numpy/issues/12518 is to pass in a struct (which could grow) with all of the important information.  The main tricky thing is to decide on:

1. Promotion (but I think multiple dispatched promotion is a good solution)
2. How to best wrap existing loops, and how to make the whole API as limited as possible for now.

### CODE REVIEW

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+
+* The ufunc itself, which may be useful for information such as the ufunc's name
+  but is also required to match up information such as the ``ufunc.signature``.
+* The DTypes involved, which are thus passed into all of the methods of the
+  ``UFuncImpl`` in some form or another.
+
+Both of these are *not* stored on the ``UFuncImpl`` itself and just like
+methods, most of the time ``UFuncImpl`` should only be used indirectly by
+calling the ufunc it is bound to.
+It is possible to bind a single ``UFuncImpl`` to multiple ufuncs.
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:335]: I think this is crucial and deserves mention early on! Here, a specific example might be good. E.g., I think `sin` and `cos` could share complete implementations, but `add` and `subtract` could not (e.g., `add` being defined for strings but not for `datetime`, and vice versa for `substract`).

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+
+
+Abstract
+--------
+
+Allowing the creation of new user-defined DTypes in NumPy also requires
+a better extensibility of the ufunc interface, which is the backbone for
+a large part of the NumPy API.
+Especially, parametric DTypes are currently severly limited, making it
+impossible to implement for a user-defined DType to register universal
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:32]: delete "to implement" here.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+function loops.
+
+This NEP proposes a new ``UFuncImpl`` struct, with an extensible C-API
+to register more feature rich loops, thus allowing these use-cases.
+It also proposes to address certain deficiencies, for example with respect
+to error reporting, by extending the signature of the core inner-loop
+functions.
+Finally, one of the most important steps in the ufunc call is finding the
+correct function for a specific set of input dtypes.  We propose to use
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:41]: correct *inner loop* function? Or just "correct inner loop"?

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+This would allow registering e.g. ``float64 + float64 -> float32`` as an explicit
+(faster) loop while indicating that it is an unsafe cast on the result array,
+which requires the user to specify ``casting="unsafe"``.
+
+The current design allows such a specialized loop (with access to the
+initially private ``setup``), from within the ``float64+float64->float64``
+implementation only.
+
+
+Bound UFuncImpl
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:317]: This section ended up confusing me no end. It is not obvious to me that it is like a bound method where the ufunc and dtypes combine to some form of `self`, unless one thinks of each `UFuncImpl` as some type of class that has `__init__(self, ufunc, dtypes)`, so that all the methods can just refer to `self.ufunc` and `self.dtypes`.

Indeed, I'm sufficiently confused that it makes me wonder whether it might not be helpful to have a mock-up of a python implementation of a `UFuncImpl` and on how one would use it for an existing function (the units case would be a good one), as well as how one would write a new `UFunc` taking perhaps an existing `UFuncImpl` (say, a `sincos` implementation). (This echoes a point made by @mattip in the original thread...)

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+However, in the above example there is no direct match, requireing a
+promotion step.
+
+**Implementation:**
+
+1. By default any UFunc has a promotion which uses the common DType of all
+   inputs and tries again.  This is well defined for most mathematical
+   functions, but can be disabled or customized if necessary.
+
+2. Users can *register* new Promoters just as they can register new UFuncImpl.
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:372]: At this point, registration of `UFuncImpl` has not really been discussed much.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+
+    def promote(ufunc, DTypes):
+        res = list(DTypes)
+        try:
+            res[1] = np.common_dtype(DTypes[1], Int64)
+        except TypeError:
+            return NotImplemented
+
+        # Could check that res[1] is actually Int64
+        return ufunc.resolve_impl(tuple(res))
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:403]: The mock-up is very useful, but would be even better if `resolve_impl` had already been introduced! (perhaps via the larger mock-up I wondered about above).

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+--------------
+
+Internally a few implementation details have to be decided. These will be
+fully opaque to the user and can be changed at a later time.
+
+This includes:
+
+* How ``CastingImpl`` lookup, and thus the decision whether a cast is possible,
+  is defined. (This is speed relevant, although mainly during a transition
+  phase where UFuncs where NEP 43 is not yet implemented).
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:588]: double where

**Review by mhvk — COMMENTED** [review]
@seberg - before anything else, a very nice write-up! Some initial comments, some just details, some perhaps too much focussed on implementation, and one real confusion, leading to the perhaps the main suggestion of some type of python mock-up.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+for casting (see also NEP 42 ``CastingImpl``).
+This gives the following function definitions:
+
+* Similar to casting, also ufuncs may need to find the correct output DType
+  or indicate that a loop is only capable of handling certain instances of
+  the involved DTypes (e.g. only native byte order).  This is handled by
+  an ``resolve_descriptors`` function (similar to ``adjust_descriptors``
+  of ``CastingImpl``)::
+
+      int resolve_descriptors(
```
[the remark this answers is not in the document]
Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst:141]: Oh, there is no intention for in-place, the naming is just confusing `in` and `out` was not supposed to refer to input and output descriptors...

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+              PyUFuncObject *ufunc, PyArray_DTypeMeta *DTypes[nargs],
+              PyArray_Descr *in_dtypes[nargs], PyArray_Descr *out_dtypes[nargs]);
+
+  setting the desired output descriptors, and possibly modifying the input
+  ones to conform with the inner-loops requirements.
+
+* Define a new structure to be passed in the inner-loop, which can be
```
[the remark this answers is not in the document]
Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst:148]: Have to think about this, thanks.  Yes, we need to wrap legacy loops in any case I think (or, hidden from the user, have a flag which causes the old inner-loop convention (depending on what is easier probably, I don't mind the indirection for loops defined outside of NumPy, and loops inside of NumPy should move ASAP as well).

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+for implementing a ``Unit`` DType.
+
+1. An abstract Unit DType with subclasses for each numeric dtype (these
+   could be dynamically created).  Such a ``Unit[Float64]("m")`` instance
+   has a well defined signature with respect to type promotion.
+2. A concrete Unit DType which parametrizes the numeric dtype making it
+   ``Unit(float64, "m")``.
+
+The difference between the two, is that in the first case a ``UFuncImpl``
+can be created which *wraps* a single existing ``UFuncImpl`` and after that
```
[the remark this answers is not in the document]
Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst:260]: Ah, this is great to here! I was seriously struggling with trying to figure out a good way to make Option 2 something that does not feel out of place...

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 11]
```diff
@@ -0,0 +1,631 @@
...
+  or indicate that a loop is only capable of handling certain instances of
+  the involved DTypes (e.g. only native byte order).  This is handled by
+  an ``resolve_descriptors`` function (similar to ``adjust_descriptors``
+  of ``CastingImpl``)::
+
+      int resolve_descriptors(
+              PyUFuncObject *ufunc, PyArray_DTypeMeta *DTypes[nargs],
+              PyArray_Descr *in_dtypes[nargs], PyArray_Descr *out_dtypes[nargs]);
+
+  setting the desired output descriptors, and possibly modifying the input
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:145]: @seberg - the "and possibly modifying the input" here was what made me think that changes were made in-place. Maybe rephrase this?

File: doc/neps/nep-0043-extensible-ufuncs.rst (hunk 4 above)
Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst:317]: Yeah, I guess you could see it that way with an `__init__(self, ufunc, dtypes)`, although, I do not think such an init has to technically be exposed to the user.  Anyway, I have to think about trying to clear this up, I was not sure I like the mental model around having an `__init__`, it somewhat implies there is a full blown (subclass-able?) class, something that I would like to avoid at least for now.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+
+   * Allocate all ouputs and potentially copies (or buffers).
+   * Finds the best iteration order, which includes information such as
+     a broadcasted stride always being 0.
+
+5. *Loop-specific setup* may include for example:
+
+   * Clearing of floating point exception flags (if necessary),
+   * Possibly allocating temporary working space,
+   * Setting (and potentially finding) the inner-loop function.  Finding
```
[the remark this answers is not in the document]
Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst:97]: True, although I think this can be part of the setup?  This is about specializations, such as contiguous loops (SIMD capable strides), or reduction loops, which can be implemented faster.  I guess the grouping here is trying to find the most practical grouping for users who want the most control possible, while not asking them to take charge of setting up the outer iteration as well.

Anyway, thanks for this and all the feedback, will try to iterate/think on it!

File: doc/neps/nep-0043-extensible-ufuncs.rst (hunk 4 above)
Comment (mhvk) [reply doc/neps/nep-0043-extensible-ufuncs.rst:317]: Agreed that the way it is exposed to a python user does not have to be decided now. My main comment here was really that I found the present text confusing (unlike the rest). Might it make sense to move this piece to a separate very preliminary discussion of how things might be exposed in python?

 Or, not as good but perhaps more practical, remove it and just have a note up front that a python implementation is not yet fully considered but that it is attempted to make sure nothing would stop it?

File: doc/neps/nep-0043-extensible-ufuncs.rst (hunk 12 above)
[the remark this answers is not in the document]
Comment (mhvk) [reply doc/neps/nep-0043-extensible-ufuncs.rst:97]: Ah, with that goal it does make sense. I think then all that's needed is to give exactly that example!

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 13]
```diff
@@ -0,0 +1,631 @@
...
+To fall back to the existing loop, we allow creation of a wrapped loop
+from an existing (bound) ``UFuncImpl``.
+
+This wrapped loop will have two additional methods:
+
+* ``view_inputs(Tuple[DType]: input_descr) -> Tuple[DType]`` replacing the
+  user input descriptors with descriptors matching the wrapped loop.
+  It must be possible to *view* the inputs as the output.
+* ``wrap_outputs(Tuple[DType]: input_descr) -> Tuple[DType]`` replacing the
+  resolved descriptors with with the desired actual loop descriptors.
```
Comment (anirudh2290) [inline doc/neps/nep-0043-extensible-ufuncs.rst:272]: I was wondering how this handles a case like np.add(Unit[Float64]("m"), Unit[Float64["km"]) ?

Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst:272]: Thanks, good point, I should explain that/give an example. For Units this uses `.view(np.float64)` in the inner loop effectively. In the outer loop, the descriptors will be set to `(Unit[Float64]("m"), Unit[Float64["m"], Unit[Float64]("m"))`, so the casting machinery has to convert the `"km"` to `"m"`.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 14]
```diff
@@ -0,0 +1,631 @@
...
+     int teardown(int return_value, PyArray_UFuncData *data);
+
+  Teardown function, may return ``-1`` if an error is set.  If an error
+  was already set previously it will be stored. Any previous error
+  will be cleared before calling ``teardown``.
+  The return value of the inner-loop is passed in to allow simple error
+  flagging (e.g. warnings) if desired.
+
+* A flag to ask NumPy to perform floating point error checking (after custom
+  setup and before user teardown).
```
[the remark this answers is not in the document]
Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst:202]: Many integer loops or loops working with Python objects only, do not really have to. And apparently floating point error checking can be a bit slow on some hardware, so I think we can create such an option.  Although, I guess it would be fair to add that later as well.

**Comment by seberg** [comment]
@hameerabbasi (and quite honestly everyone). While I will iterate on it in the next days, I think the big picture stuff is there and should be fine to read, so please do whenever you have time, independently of me updating it soon.  It will require more iterations in any case, delaying the start of the discussion just means that my iterations are more annoying, slower and less thorough.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 15]
```diff
@@ -0,0 +1,631 @@
...
+
+* The *optional* teardown function::
+
+     int teardown(int return_value, PyArray_UFuncData *data);
+
+  Teardown function, may return ``-1`` if an error is set.  If an error
+  was already set previously it will be stored. Any previous error
+  will be cleared before calling ``teardown``.
+  The return value of the inner-loop is passed in to allow simple error
+  flagging (e.g. warnings) if desired.
```
Comment (seberg) [inline doc/neps/nep-0043-extensible-ufuncs.rst:199]: Note for those who already read this. I think I will modify this. I am starting to be convinced that `teardown` does not need to do any error-related work, and that it may be an anti-pattern.  Floating point errors are special, since they require this.  As such, we could allow an error return, but should discourage it, since its much like deleting a Python object.

File: doc/neps/nep-0043-extensible-ufuncs.rst (hunk 3 above)
Comment (hameerabbasi) [reply doc/neps/nep-0043-extensible-ufuncs.rst:41]: How about "correct ufunc implementation"?

File: doc/neps/nep-0043-extensible-ufuncs.rst (hunk 4 above)
Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst]: I have expanded/refactored this section, and more importantly tried to add a super-highlevel-get-the-feel for things section at the start, which happens to include `resolve_impl` (I don't know how much that helps, and I guess you are the wrong person to ask by now).

In any case, this section is restructured, and I tried to point out that it is a bit nitty-gritty for the interested reader (I am unsure about it, but it helped me be clear about who has what information)

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 16]
```diff
@@ -0,0 +1,836 @@
...
+    add_impl = np.add.resolve_impl(type(arr1.dtype), type(arr2.dtype))
+
+This NEP strives to define the minimal set of API necessary to achieve
+the above lists of steps and make it fully extensible for authors of both
+custom ufuncs and DTypes.
+There are two distinct API design decisions as part of this NEP:
+
+1. For the above to work, we must also define::
+
+       np.positive.resolve_impl(MyDateTime)
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:113]: Maybe just `MyDType` here and below? Or  in the sentence above, "1. For the above to work for, e.g., a `MyDateTime` DType, we must also define"

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 17]
```diff
@@ -0,0 +1,836 @@
...
+Introduction to General Concepts
+""""""""""""""""""""""""""""""""
+
+Universal functions as described above are much like a Python method
+defined on the DType of the array when considering a ufunc with only
+a single input::
+
+    res = np.positive(arr)
+
+could be implemented (conceptional) as:
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:73]: conceptional*ly*

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 18]
```diff
@@ -0,0 +1,836 @@
...
+
+is not part of this NEP, as it is not central to the proposal.
+
+As mentioned above, in general NumPy universal functions can take many
+inputs, requiring looking up the implementation by considering all of them
+making NumPy ufunc "multi-methods" with respect to the input *DTypes*::
+
+    add_impl = np.add.resolve_impl(type(arr1.dtype), type(arr2.dtype))
+
+This NEP strives to define the minimal set of API necessary to achieve
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:106]: delete "set of" (i.e., just "minimal API")

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 19]
```diff
@@ -0,0 +1,836 @@
...
+       np.positive.resolve_impl(MyDateTime)
+
+   In the case of multiple inputs, often *promotion* must occur:
+   ``float32 + float64`` is actually handled by the implementation for
+   ``float64 + float64``.  NumPy automatically converts (casts) the first
+   argument.
+
+   This is described in the "Promotion and Dispatching" section and
+   step 2 in the list included in the next section
+   (The Steps involved in a UFunc call).
```
Comment (mhvk) [inline doc/neps/nep-0043-extensible-ufuncs.rst:122]: Maybe just link "next section" in rst?

**Review by mhvk — COMMENTED** [review]
@seberg - I like your example up front! Just a few super-nitpicky comments as I was reading through.

**Review by hameerabbasi — COMMENTED** [review]
I've only read this partway through as of yet.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 20]
```diff
@@ -0,0 +1,631 @@
...
+3. *DType Adaptation:*
+
+   * The step has to perform no special work for non-parametric dtypes.
+   * For example, if a loop adds two strings, it is necessary to define the
+     correct output (and possibly input) dtypes.  ``S5 + S4 -> S9``, while
+     an ``upper`` function has the signature ``S5 -> S5``.
+
+4. Preparing the actual iteration. This step is largely handled by ``NpyIter`` (the iterator).
+
+   * Allocate all ouputs and potentially copies (or buffers).
```
[the remark this answers is not in the document]
Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst]: I deleted it. For example its the implementation for overlap protection and also for very small arrays more efficient (dunno if we use it for anything else but scalars).

But: It just doesn't matter here.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 21]
```diff
@@ -0,0 +1,895 @@
...
+This NEP requires to partially represent this implementation explicitly::
+
+    positive_impl = np.positive.resolve_impl(type(arr.dtype))
+
+Although, this ``positive_impl`` and ``resolve_impl`` is required for certain
+functionality, the following code::
+
+    res = positive_impl(arr)
+
+is not part of this NEP, as it is not central to the proposal.
```
Comment (hameerabbasi) [inline doc/neps/nep-0043-extensible-ufuncs.rst:98]: Possibly move to a non-goals section.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 22]
```diff
@@ -0,0 +1,895 @@
...
+Universal functions as described above are much like a Python method
+defined on the DType of the array when considering a ufunc with only
+a single input::
+
+    res = np.positive(arr)
+
+could be implemented (conceptionally) as:
+
+    positive_impl = arr.dtype.positive
+    res = positive_impl(arr)
```
Comment (hameerabbasi) [inline doc/neps/nep-0043-extensible-ufuncs.rst:76]: If this isn't what happens, and not something that the NEP will modify (as you say below), then I'd remove it completely or replace it by a more accurate statement.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 23]
```diff
@@ -0,0 +1,895 @@
...
+Introduction to General Concepts
+""""""""""""""""""""""""""""""""
+
+Universal functions as described above are much like a Python method
+defined on the DType of the array when considering a ufunc with only
+a single input::
+
+    res = np.positive(arr)
+
+could be implemented (conceptionally) as:
```
Comment (hameerabbasi) [inline doc/neps/nep-0043-extensible-ufuncs.rst:73]:
```suggestion
could be implemented (conceptually) as:
```

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 24]
```diff
@@ -0,0 +1,895 @@
...
+The following sections go into more details, and are seperated into the
+two main topics of *promotion and dispatching* and the further C-API
+provided to the user for the ufunc execution.
+
+
+UFuncImpl Registration
+""""""""""""""""""""""
+
+*TODO:* we need to briefly mention registration, even if the details of
+how to register it are in the specs or even later!
```
Comment (hameerabbasi) [inline doc/neps/nep-0043-extensible-ufuncs.rst:214]: Yes, we should. API decisions should come before implementation details. 😄

**Review by hameerabbasi — COMMENTED** [review]
In general, I'd front-load the use-cases for this NEP instead of front-loading the implementation. I realize that separating the two is hard; and also that as developers who implement things, we tend to focus more on implementation.

However, this makes the end-user zone out if they don't understand, and we lose their interest. [See this discussion for details](https://mail.python.org/pipermail/numpy-discussion/2020-August/080942.html).

File: doc/neps/nep-0043-extensible-ufuncs.rst (hunk 22 above)
Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst]: @hameerabbasi, I am trying to give the reader an idea/concept of what a multi-method is (and how it comes to into play with ufuncs). Starting with normal methods, seemed like a fair way to do that to me?  Maybe the sentence afterwards can say that this is not how it can be done in NumPy for reasons 1 and 2.  (Although, I have only semi appetite to discuss the possibility of an `__array_ufunc__` like approach, unless someone insists on it being listed as a rejected alternative.)

Comment (hameerabbasi) [reply doc/neps/nep-0043-extensible-ufuncs.rst]: How about the following:

```python
# find_actual_implementation is what we want to achieve for custom dtypes
# and ufuncs
positive_impl = find_actual_implementation(np.positive, arr.dtype)
res = positive_impl(arr)
```

I just don't want to give users who gloss over this the perception that `positive` is a `dtype` slot. It's best to choose an example which doesn't require you to explain that.

Comment (seberg) [reply doc/neps/nep-0043-extensible-ufuncs.rst]: Thanks, I wanted the concept of "method" in there I guess, but maybe that is just not necessary.

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 25]
```diff
@@ -0,0 +1,1261 @@
...
+    - NEP 43 (this document) describes the new design's API for universal functions.
+
+
+******************************************************************************
+Abstract
+******************************************************************************
+
+The previous NEP 42 document proposes the creation of new DTypes which can
+be defined by users outside of NumPy itself.
+While implementing NEP 42 will users to create arrays with a custom dtype
```
Comment (dhirschfeld) [inline doc/neps/nep-0043-extensible-ufuncs.rst:33]:
```suggestion
While implementing NEP 42 will enable users to create arrays with a custom dtype
```

File: doc/neps/nep-0043-extensible-ufuncs.rst [hunk 26]
```diff
@@ -0,0 +1,1261 @@
...
+            
+            # The operation is always "safe" casting (most ufuncs are)
+            return (given_descrs[0], given_descrs[1], context.DTypes[2]()), "safe"
+
+        def strided_loop(context, n, data, strides):
+            # n: Number of 
+            # data: Pointers to the array data.
+            # strides: strides to iterate all elements
+            num_chars1 = context.descriptors[0].itemsize
+            num_chars2 = context.descriptors[0].itemsize
```
Comment (mattip) [inline doc/neps/nep-0043-extensible-ufuncs.rst:257]: I think you need dimension as well as strides. If dimensions are in the context, then strides could be as well. Also `# n: Number of` seems like a fragment.

**Review by mattip — COMMENTED** [review]
Here is a link to the renedered document https://766-5652112-gh.circle-artifacts.com/0/doc/neps/_build/html/nep-0043-extensible-ufuncs.html. 

I think we should merge this soon, there is an open typo suggestion and I made a suggestion around `strided_loop`, but feel free to ignore and we can iterate after this is merged.

**Review by rossbar — COMMENTED** [review]
This is obviously quite the undertaking - thanks @seberg for taking the time to compile all this information, and for taking a lot of care in the communication of the concepts!

The only high-level comment that I have is that I would've found it useful if the "steps in a ufunc call" were higher up in the document, as this provides a nice scaffolding for all of the subsequent pieces that are addressed individually. It's not a show-stopper as there are forward references to that section in some places.

**Comment by seberg** [comment]
Thanks for the review especially Matti, Ross and of course Marten! Going to put this in so that we have an easy to read draft, but of course this is will not be the last iteration.

**Comment by guidocalvano** [comment]
What is the state of this? The draft of this is now in master, but it is not considered an accepted NEP yet?

https://numpy.org/neps/nep-0043-extensible-ufuncs.html

**Comment by seberg** [comment]
I am not sure we ever marked it as accepted (which we should do with the note below, probably).

The API probably moved a bit since then, but yes, I would like most of this to be available in NumPy 2.0.  We are using the API internally quite a bit.  I will note that there is still a chance of doing some changes, or omitting some functions to allow easier changes in the near future.

I am still considering one larger change mainly:  Re-adding the static `void *` pointer as part of the `context` struct as well as moving the new `auxdata` into the `context` struct.  If you have any thoughts, please let us know, this is the last few weeks were changes are trivial.
(The new API is designed to be able to deprecate older versions, but especially for the ufunc inner-loop signature that seems tedious.)

The new API is available for experimentation for a long time via `extermintal_dtype_api.h`.

**Comment by guidocalvano** [comment]
Is there a python api already? I don't expect to need c code.

Some more context should you be interested:
What I want to do is implement two things: decimal types and multivectors from geometric algebra. I expect both to be implementable fairly easily and quickly by a python ufunc: i.e. you create a dtype that imposes a size on an element, and then you can create a view of a different dtype, to access the data. Then you can do your math by operating on the view and return whatever the operator requires. Decimal add can be done by simply viewing the values as ints and adding them. Multiplication can be done by promoting to a bigger int data type, doing the multiplication, then shifting values two times the numbers behind the decimal point. Multivectors are a bit more complicated... it would take at least half an hour to explain the concept... but well worth watching a few videos on.

**Comment by seberg** [comment]
I have no plans for Python API for the foreseeable future, it would need a higher level hook to do well probably.  Yes, you could add that, but no it doesn't exist.

**Comment by mhvk** [comment]
@seberg - I think an ability to pass on arbitrary information to the inner loop remains a good idea. E.g., using old-style I have used it to create "ufunc chains" where the inner loop executes other inner loops in order - for speed-ups comparable to numexpr without having to describe operations in strings (see https://github.com/mhvk/chain_ufunc). And of course, you saw that it was somewhat handy for the gufunc fft/ifft routines in #25536...

But the main problem in commenting is that while I reviewed quite a bit, I haven't yet *used* the new API, so I don't really know how it would work in practice... So take this with a pinch of salt!
